engine/execution/state_synchronization/state_diff_storer.go

```diff
@@ -0,0 +1,179 @@
+package state_synchronization
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/network"
+)
+
+const MAX_BLOCK_SIZE = 1e6
+
+type ExecutionStateDiff struct {
+	Collections        []*flow.Collection
+	Events             []*flow.Event
+	TrieUpdate         []*ledger.TrieUpdate
+	TransactionResults []*flow.TransactionResult
+}
+
+type StateDiffStorer struct {
+	blockWriter *BlockWriter
+	codec       encoding.Codec
+	compressor  network.Compressor
+}
+
+func NewStateDiffStorer(
+	codec encoding.Codec,
+	compressor network.Compressor,
+	bstore blockstore.Blockstore,
+) (*StateDiffStorer, error) {
+	bw := &BlockWriter{
+		maxBlockSize: MAX_BLOCK_SIZE,
+		bstore:       bstore,
+	}
+	return &StateDiffStorer{bw, codec, compressor}, nil
+}
+
+func (s *StateDiffStorer) writeBlocks(v interface{}) ([]cid.Cid, error) {
+	comp, err := s.compressor.NewWriter(s.blockWriter)
+	if err != nil {
+		return nil, err
+	}
+	enc := s.codec.NewEncoder(comp)
+	if err := enc.Encode(v); err != nil {
+		return nil, err
+	}
+	if err := comp.Close(); err != nil {
+		return nil, err
+	}
+	if err := s.blockWriter.Flush(); err != nil {
+		return nil, err
+	}
+	cids := s.blockWriter.GetWrittenCids()
+	s.blockWriter.Reset()
+	return cids, nil
+}
+
+func (s *StateDiffStorer) Store(sd *ExecutionStateDiff) (cid.Cid, error) {
+	if cids, err := s.writeBlocks(sd); err != nil {
+		return cid.Undef, fmt.Errorf("failed to write state diff blocks: %w", err)
+	} else {
+		for {
+			if len(cids) == 1 {
+				return cids[0], nil
+			}
+
+			if cids, err = s.writeBlocks(cids); err != nil {
+				return cid.Undef, fmt.Errorf("failed to write cid blocks: %w", err)
+			}
+		}
+	}
+}
+
+func (s *StateDiffStorer) readBlocks(cids []cid.Cid, v interface{}) error {
+	buf := &bytes.Buffer{}
+	comp, err := s.compressor.NewReader(buf)
+	if err != nil {
+		return err
+	}
+	dec := s.codec.NewDecoder(comp)
+	for _, c := range cids {
+		block, err := s.blockWriter.bstore.Get(c)
+		if err != nil {
+			return fmt.Errorf("failed to get block %v from blockstore: %w", c, err)
+		}
+		_, _ = buf.Write(block.RawData()) // never returns error
+	}
+	return dec.Decode(v)
+}
+
+func (s *StateDiffStorer) Load(c cid.Cid) (*ExecutionStateDiff, error) {
+	cids := []cid.Cid{c}
+	for {
+		var sd ExecutionStateDiff
+		if err := s.readBlocks(cids, &sd); err == nil {
+			return &sd, nil
+		} else {
+			if err := s.readBlocks(cids, &cids); err != nil {
```

Comment:
how common are state diffs >1MB? performing tons of these double read/decodes will waste CPU/IO resources on the ENs. if it's common enough, we might consider adding a header to the data so we can differentiate the type before attempting to decode it. that would probably require creating our own block type that includes the type as a separate field outside of the data slice since it shouldn't be included in the hash

Reply:
They are very uncommon, more than 99% of the time the first if will succeed.

Reply:
if these are very uncommon, I'm fine punting on this as a future optimization

Reply:
> Also, I haven't checked this but I would assume CBOR already adds headers to the data to differentiate the type. Perhaps @fxamacker can confirm this?

Hi @smnzhu,  yes, CBOR uses a generic "CBOR head" to identify CBOR data types ([major types](https://www.rfc-editor.org/rfc/rfc8949.html#name-major-types)) of encoded values.

In addition, the encoded CBOR head+value (data item) can be enclosed by a [CBOR tag](https://www.rfc-editor.org/rfc/rfc8949.html#name-tagging-of-items) to give the data item additional semantics.  For example, Cadence uses [CBOR tags to identify Cadence-specific data types](https://github.com/onflow/cadence/blob/master/runtime/interpreter/encode.go#L95-L207).

CBOR tags can be used here to indicate Flow-specific types, but it will affect data size (typically a few bytes) and possibly performance.

Reply:
2ad73ab @peterargue Implemented a generic solution for now to support all encoding implementations, but eventually we will probably settle on a single encoding type (probably CBOR), and can use native solution such as CBOR tag, as mentioned above.